### PR TITLE
Added level loading and scene manager

### PR DESCRIPTION
**Pull request by Bram-Boris**
# Description

The GameController now contains code which will load a beautifully designed level from a JSON file. I does so by using the JSON adapter from the engine to create a level model. This level model can be converted to entities by the SceneManager. The SceneManager is also able the destroy the whole scene.

At last, some changes were made to the EntityFactory because the engine now contains alpha (transparency).

## Screenshot(s)
image

# How can this be tested?

You can start off by compiling the game (and thus [engine](https://github.com/Brick-Studios/BrickEngine/pull/39)) and check if the level looks like the screenshot. There are no tests for the SceneManager because that is too complicated but you could tinker around with the JSON level files.

- [ ] The game compiles and looks like the screenshot

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Included 1 or more screenshots
- [x] Code is const correct

### CODE REVIEW

File: include/exceptions/notEnoughPlayerSpawnsException.hpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#ifndef FILE_NOTENOUGHPLAYERSPAWNEXCEPTION_HPP
+#define FILE_NOTENOUGHPLAYERSPAWNEXCEPTION_HPP
+
+#include <iostream>
+#include <exception>
+
+class NotEnoughPlayerSpawnsException : public std::exception {
```
Comment (JanSchollaert) [inline include/exceptions/notEnoughPlayerSpawnsException.hpp]: notEnoughPlayerSpawnsException.hpp 
wrong naming (snake_case)

File: include/exceptions/sizeMismatchException.hpp [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#ifndef FILE_SIZEMISMATCHEXCEPTION_HPP
+#define FILE_SIZEMISMATCHEXCEPTION_HPP
+
+#include <iostream>
+#include <exception>
+
+class SizeMismatchException : public std::exception {
```
Comment (JanSchollaert) [inline include/exceptions/sizeMismatchException.hpp]: sizeMismatchException.hpp
wrong naming (snake_case)

File: include/exceptions/sizeMismatchException.hpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+private:
+  std::string m_msg;
+public:
+  SizeMismatchException() : m_msg(std::string("Level is not 16:9 and cannot be transformed to game size")){}
+
+  virtual const char* what() const throw() {
```
Comment (JanSchollaert) [inline include/exceptions/sizeMismatchException.hpp]: Maybe a comment about what what() does? Its kind of a strange name

File: lib/controllers/game_controller.cpp [hunk 4]
```diff
@@ -46,12 +55,85 @@ void GameController::createSystems() {
...
-    entityFactory->createImage("backgrounds/forest_watermarked.jpg", SCREEN_WIDTH / 2, SCREEN_HEIGHT / 2, SCREEN_WIDTH, SCREEN_HEIGHT, Layers::Background);
-    entityFactory->createPlatform(1200, 680, 400, 10);
-    entityFactory->createPlatform(400, 680, 400, 10);
-    entityFactory->createPlatform(800, 510, 800, 10);
+    entityFactory->createPanda(0, 0, 1);
+    entityFactory->createGorilla(0, 0, 2);
+
+    Json box_json = Json("assets/levels/level2.json", true);
+    auto level = loadLevel(box_json);
+    scene_manager->loadLevel(level.get());
```
Comment (JanSchollaert) [inline lib/controllers/game_controller.cpp]: Is this just for testing? If not it shouldnt be in createTestEntities()

**Review by JanSchollaert — DISMISSED** [review]
Some small naming change requests

File: include/exceptions/sizeMismatchException.hpp (hunk 3 above)
Comment (Peter-JanGootzen) [reply include/exceptions/sizeMismatchException.hpp]: TOP KEK :rofl:

File: include/controllers/game_controller.hpp [hunk 5]
```diff
@@ -25,6 +26,7 @@ class GameController {
     std::shared_ptr<EntityManager> entityManager;
     std::shared_ptr<EntityFactory> entityFactory;
     std::shared_ptr<CollisionDetector> collisionDetector;
+    std::shared_ptr<SceneManager> scene_manager;
```
Comment (Peter-JanGootzen) [inline include/controllers/game_controller.hpp]: This can be a unique_ptr

File: lib/scenes/scene_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+#include "level/level.hpp"
+#include "entities/layers.hpp"
+#include "level/solid.hpp"
+#include "level/player_spawn.hpp"
+#include "brickengine/components/player_component.hpp"
+#include "brickengine/components/transform_component.hpp"
+
+SceneManager::SceneManager(std::shared_ptr<EntityFactory> entity_factory, std::shared_ptr<EntityManager> entity_manager, BrickEngine* engine) : entity_factory(entity_factory), entity_manager(entity_manager), engine(engine) {};
+
+void SceneManager::loadLevel(Level* level) {
```
Comment (Peter-JanGootzen) [inline lib/scenes/scene_manager.cpp]: level should be a reference

File: lib/controllers/game_controller.cpp [hunk 7]
```diff
@@ -46,12 +53,12 @@ void GameController::createSystems() {
...
-    entityFactory->createGorilla(1000, 200, 2);
-    entityFactory->createImage("backgrounds/forest_watermarked.jpg", SCREEN_WIDTH / 2, SCREEN_HEIGHT / 2, SCREEN_WIDTH, SCREEN_HEIGHT, Layers::Background);
-    entityFactory->createPlatform(1200, 680, 400, 10);
-    entityFactory->createPlatform(400, 680, 400, 10);
-    entityFactory->createPlatform(800, 510, 800, 10);
+    entityFactory->createPanda(0, 0, 1);
+    entityFactory->createGorilla(0, 0, 2);
+
+    Json level_json = Json("assets/levels/level2.json", true);
+    auto level = std::unique_ptr<Level>(new Level(level_json, SCREEN_WIDTH, SCREEN_HEIGHT));
```
Comment (Peter-JanGootzen) [inline lib/controllers/game_controller.cpp]: No need to allocate this on heap, just make it a local variable and then pass it to the loadLevel function.

**Review by Peter-JanGootzen — CHANGES_REQUESTED** [review]

**Review by Peter-JanGootzen — APPROVED** [review]
I was thinking of suggesting to use the visitor pattern in the scene_manager, because you now have a nice inheritance tree with scene, level (and soon menu) but you have to call the correct function instead of a generic one. But even though this would _cool and correct_, I can't think of any scenario in a game where you wouldn't know whether you currently want to load a level or a menu. If you don't know a scenario either, then I'll approve this PR.

**Comment by Bram-Boris** [comment]
I also can not think of any scenario :wink:

**Review by JanSchollaert — APPROVED** [review]
Only thing is that the platforms are very dark and the gorilla as well which makes it harder to see everything clearly. But this is easily fixed by using different textures.

**Pull request merged**